README.md
# body-gamepad
Emulate arrow keys and two buttons with your arms and legs.


CHANGELOG.rst
**********
Change log
**********

0.0.1 (2023-04-15)
==================
* This is the initial version.


bodygamepad/bodypose.py
"""Recognize the body pose of the player."""

import collections
import enum
import math
import os
import pathlib
from typing import List, Final, Mapping, Callable, MutableMapping, Tuple, Any

import cv2
import numpy as np
import tensorflow as tf
import tensorflow_hub as hub
from icontract import require


class KeypointLabel(enum.Enum):
    """Map keypoints names to the indices in the network output."""

    NOSE = 0
    LEFT_EYE = 1
    RIGHT_EYE = 2
    LEFT_EAR = 3
    RIGHT_EAR = 4
    LEFT_SHOULDER = 5
    RIGHT_SHOULDER = 6
    LEFT_ELBOW = 7
    RIGHT_ELBOW = 8
    LEFT_WRIST = 9
    RIGHT_WRIST = 10
    LEFT_HIP = 11
    RIGHT_HIP = 12
    LEFT_KNEE = 13
    RIGHT_KNEE = 14
    LEFT_ANKLE = 15
    RIGHT_ANKLE = 16


#: Map indices of the network output to keypoint labels
KEYPOINT_INDEX_TO_LABEL = {literal.value: literal for literal in KeypointLabel}


class Keypoint:
    """Represent a single detection of a keypoint in an image."""

    x: float
    y: float
    confidence: float

    @require(lambda confidence: 0 <= confidence <= 1)
    def __init__(self, x: float, y: float, confidence: float) -> None:
        """
        Initialize with the given values.

        :param x: X-coordinate in the image rescaled to [0, 1] x [0, 1]
        :param y: Y-coordinate in the image rescaled to [0, 1] x [0, 1]
        :param confidence: in the range [0,1] of the keypoint detection
        """
        self.x = x
        self.y = y
        self.confidence = confidence


class Detection:
    """Represent a detection of a person in an image."""

    #: Keypoints of the pose
    keypoints: Final[Mapping[KeypointLabel, Keypoint]]

    #: Score of the person detection.
    #:
    #: .. note::
    #:
    #:     This score is the score of the *person* detection, not of the individual
    #:     joints. For the score of the individual joints,
    #:     see :py:attr:`Keypoint.confidence`

    @require(lambda score: 0 <= score <= 1)
    def __init__(
        self,
        keypoints: Mapping[KeypointLabel, Keypoint],
        score: float,
    ) -> None:
        """Initialize with the given values."""
        self.keypoints = keypoints
        self.score = score


def load_empty_detector() -> Callable[[cv2.Mat], List[Detection]]:
    """
    Create a detector that always returns an empty result.

    This is used for debugging.
    """

    def apply_model(img: cv2.Mat) -> List[Detection]:
        return []

    return apply_model


# noinspection SpellCheckingInspection
def _load_tf_model(path: pathlib.Path) -> Any:
    """
    Load the TF model from disk.

    This function is an adaption of ``tensorflow_hub.load(.)``
    """
    module_path = str(path)

    is_hub_module_v1 = tf.io.gfile.exists(
        hub.native_module.get_module_proto_path(module_path)
    )

    saved_model_path = os.path.join(
        tf.compat.as_bytes(module_path),
        tf.compat.as_bytes(tf.saved_model.SAVED_MODEL_FILENAME_PB),
    )
    saved_model_pbtxt_path = os.path.join(
        tf.compat.as_bytes(module_path),
        tf.compat.as_bytes(tf.saved_model.SAVED_MODEL_FILENAME_PBTXT),
    )
    if not tf.io.gfile.exists(saved_model_path) and not tf.io.gfile.exists(
        saved_model_pbtxt_path
    ):
        raise ValueError(
            f"Trying to load a model of incompatible/unknown type. "
            f"{module_path} contains neither {tf.saved_model.SAVED_MODEL_FILENAME_PB} "
            f"nor {tf.saved_model.SAVED_MODEL_FILENAME_PBTXT}."
        )

    obj = tf.compat.v1.saved_model.load_v2(module_path)
    obj._is_hub_module_v1 = is_hub_module_v1  # pylint: disable=protected-access
    return obj


@require(lambda path: path.exists() and path.is_dir())
def load_detector(path: pathlib.Path) -> Callable[[cv2.Mat], List[Detection]]:
    """
    Load the model and return the function which you can readily use on images.

    :param path: to the model directory
    :return: detector function to be applied on images
    """
    model = _load_tf_model(path)

    movenet = model.signatures["serving_default"]

    # If a detection has a score below this threshold, it will be ignored.
    detection_score_threshold = 0.2

    # If a keypoint has a confidence below this threshold, it will be ignored.
    keypoint_confidence_threshold = 0.2

    def apply_model(img: cv2.Mat) -> List[Detection]:
        # NOTE (mristin, 2023-02-26):
        # Vaguely based on:
        # * https://www.tensorflow.org/hub/tutorials/movenet,
        # * https://www.section.io/engineering-education/multi-person-pose-estimator-with-python/,
        # * https://analyticsindiamag.com/how-to-do-pose-estimation-with-movenet/ and
        # * https://github.com/geaxgx/openvino_movenet_multipose/blob/main/MovenetMPOpenvino.py

        # Both height and width need to be multiple of 32,
        # height to width ratio should resemble the original image, and
        # the larger side should be made to 256 pixels.
        #
        # Example: 720x1280 should be resized to 160x256.

        height, width, _ = img.shape

        input_size = 256

        if height > width:
            new_height = input_size
            # fmt: off
            new_width = int(
                (float(width) * float(new_height) / float(height)) // 32
            ) * 32
            # fmt: on
        else:
            new_width = input_size
            # fmt: off
            new_height = int(
                (float(height) * float(new_width) / float(width)) // 32
            ) * 32
            # fmt: on

        if new_height != height or new_width != width:
            resized = cv2.resize(img, (new_width, new_height))
        else:
            resized = img

        tf_input_img = tf.cast(
            tf.image.resize_with_pad(
                image=tf.expand_dims(resized, axis=0),
                target_height=new_height,
                target_width=new_width,
            ),
            dtype=tf.int32,
        )

        inference = movenet(tf_input_img)
        output_as_tensor = inference["output_0"]
        assert output_as_tensor.shape == (1, 6, 56)

        output = np.squeeze(output_as_tensor)
        assert output.shape == (6, 56)

        detections = []  # type: List[Detection]

        for detection_i in range(6):
            kps = output[detection_i][:51].reshape(17, -1)
            bbox = output[detection_i][51:55].reshape(2, 2)
            score = output[detection_i][55]

            if score < detection_score_threshold:
                continue

            assert kps.shape == (17, 3)
            assert bbox.shape == (2, 2)

            kps_xy = kps[:, [1, 0]]
            kps_confidence = kps[:, 2]

            assert kps_xy.shape == (17, 2)
            assert kps_confidence.shape == (17,)

            keypoints = (
                collections.OrderedDict()
            )  # type: MutableMapping[KeypointLabel, Keypoint]

            for keypoint_i in range(17):
                label = KEYPOINT_INDEX_TO_LABEL[keypoint_i]
                kp_x, kp_y = kps_xy[keypoint_i, :]
                kp_confidence = kps_confidence[keypoint_i]

                if kp_confidence < keypoint_confidence_threshold:
                    continue

                assert label not in keypoints
                keypoints[label] = Keypoint(kp_x, kp_y, kp_confidence)

            detection = Detection(keypoints, score)

            detections.append(detection)

        return detections

    return apply_model


@require(
    lambda hip, knee, ankle: hip[1] > ankle[1] and knee[1] > ankle[1],
    "Coordinate origin in the bottom-left of the image, not in the top-left",
)
def compute_knee_angle(
    hip: Tuple[float, float], knee: Tuple[float, float], ankle: Tuple[float, float]
) -> float:
    """
    Compute the angle between the knee and the other two points.

    Going right means the negative angle:
    >>> round(compute_knee_angle((0, 2), (1, 1), (0, 0)))
    -90

    Squatting means smaller angle when going to the right:
    >>> round(compute_knee_angle((0, 0.5), (1, 0.25), (0, 0)))
    -28

    Going left means the positive angle:
    >>> round(compute_knee_angle((0, 2), (-1, 1), (0, 0)))
    90

    Squatting means smaller angle also to the left:
    >>> round(compute_knee_angle((0, 0.5), (-1, 0.25), (0, 0)))
    28

    Going straight means 180:
    >>> round(compute_knee_angle((0, 2), (0, 1), (0, 0)))
    180

    Some observations regarding the body pose:

    * It seems that going left/right is indicated when the angle goes below 150 degrees.
    * The bent of 120 degrees is already a stretch for the body.
    * 90 degrees is almost impossible.
    """
    # See: https://stackoverflow.com/a/31334882/1600678
    rads = math.atan2(hip[1] - knee[1], hip[0] - knee[0]) - math.atan2(
        ankle[1] - knee[1], ankle[0] - knee[0]
    )

    degrees = rads / math.pi * 180.0
    if degrees > 180:
        # NOTE (mristin, 2023-02-26):
        # We transform the angle so that we can simply compare the sign for
        # the direction and the magnitude for the speed.
        degrees = -(360 - degrees)

    return degrees


continuous_integration/check_init_and_setup_coincide.py
#!/usr/bin/env python3

"""Emulate arrow keys and two buttons with your arms and legs."""

import os
import pathlib
import subprocess
import sys
from typing import Optional, Dict

import bodygamepad


def main() -> int:
    """Execute the main routine."""
    repo_root = pathlib.Path(os.path.realpath(__file__)).parent.parent

    setup_py_pth = repo_root / "setup.py"
    if not setup_py_pth.exists():
        raise RuntimeError(f"Could not find_our_type the setup.py: {setup_py_pth}")

    success = True

    ##
    # Check basic fields
    ##

    setup_py_map = dict()  # type: Dict[str, str]

    fields = ["version", "author", "license", "description"]
    for field in fields:
        out = subprocess.check_output(
            [sys.executable, str(repo_root / "setup.py"), f"--{field}"],
            encoding="utf-8",
        ).strip()

        setup_py_map[field] = out

    if setup_py_map["version"] != bodygamepad.__version__:
        print(
            f"The version in the setup.py is {setup_py_map['version']}, "
            f"while the version in bodygamepad/__init__.py is: "
            f"{bodygamepad.__version__}",
            file=sys.stderr,
        )
        success = False

    if setup_py_map["author"] != bodygamepad.__author__:
        print(
            f"The author in the setup.py is {setup_py_map['author']}, "
            f"while the author in bodygamepad/__init__.py is: "
            f"{bodygamepad.__author__}",
            file=sys.stderr,
        )
        success = False

    if setup_py_map["license"] != bodygamepad.__license__:
        print(
            f"The license in the setup.py is {setup_py_map['license']}, "
            f"while the license in bodygamepad/__init__.py is: "
            f"{bodygamepad.__license__}",
            file=sys.stderr,
        )
        success = False

    if setup_py_map["description"] != bodygamepad.__doc__:
        print(
            f"The description in the setup.py is {setup_py_map['description']}, "
            f"while the description in bodygamepad/__init__.py is: "
            f"{bodygamepad.__doc__}",
            file=sys.stderr,
        )
        success = False

    ##
    # Classifiers need special attention as there are multiple.
    ##

    # This is the map from the distribution to expected status in __init__.py.
    status_map = {
        "Development Status :: 1 - Planning": "Planning",
        "Development Status :: 2 - Pre-Alpha": "Pre-Alpha",
        "Development Status :: 3 - Alpha": "Alpha",
        "Development Status :: 4 - Beta": "Beta",
        "Development Status :: 5 - Production/Stable": "Production/Stable",
        "Development Status :: 6 - Mature": "Mature",
        "Development Status :: 7 - Inactive": "Inactive",
    }

    classifiers = (
        subprocess.check_output(
            [sys.executable, str(setup_py_pth), "--classifiers"], encoding="utf-8"
        )
        .strip()
        .splitlines()
    )

    status_classifier = None  # type: Optional[str]
    for classifier in classifiers:
        if classifier in status_map:
            status_classifier = classifier
            break

    if status_classifier is None:
        print(
            "Expected a status classifier in setup.py "
            "(e.g., 'Development Status :: 3 - Alpha'), but found none.",
            file=sys.stderr,
        )
        success = False
    else:
        expected_status_in_init = status_map[status_classifier]

        if expected_status_in_init != bodygamepad.__status__:
            print(
                f"Expected status {expected_status_in_init} "
                f"according to setup.py in bodygamepad/__init__.py, "
                f"but found: {bodygamepad.__status__}"
            )
            success = False

    if not success:
        return -1

    return 0


if __name__ == "__main__":
    sys.exit(main())


continuous_integration/precommit.py
#!/usr/bin/env python3

"""Run pre-commit checks on the repository."""

import argparse
import enum
import os
import pathlib
import shlex
import subprocess
import sys
from typing import Optional, Mapping, Sequence


# pylint: disable=unnecessary-comprehension


class Step(enum.Enum):
    """Enumerate different pre-commit steps."""

    REFORMAT = "reformat"
    MYPY = "mypy"
    PYLINT = "pylint"
    TEST = "test"
    DOCTEST = "doctest"
    CHECK_INIT_AND_SETUP_COINCIDE = "check-init-and-setup-coincide"
    CHECK_HELP_IN_README = "check-help-in-readme"


def call_and_report(
    verb: str,
    cmd: Sequence[str],
    cwd: Optional[pathlib.Path] = None,
    env: Optional[Mapping[str, str]] = None,
) -> int:
    """
    Wrap a subprocess call with the reporting to STDERR if it failed.

    Return 1 if there is an error and 0 otherwise.
    """
    cmd_str = " ".join(shlex.quote(part) for part in cmd)

    if cwd is not None:
        print(f"Executing from {cwd}: {cmd_str}")
    else:
        print(f"Executing: {cmd_str}")

    exit_code = subprocess.call(cmd, cwd=str(cwd) if cwd is not None else None, env=env)

    if exit_code != 0:
        print(
            f"Failed to {verb} with exit code {exit_code}: {cmd_str}", file=sys.stderr
        )

    return exit_code


def main() -> int:
    """Execute entry_point routine."""
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--overwrite",
        help="Try to automatically fix the offending files (e.g., by re-formatting).",
        action="store_true",
    )
    parser.add_argument(
        "--select",
        help=(
            "If set, only the selected steps are executed. "
            "This is practical if some of the steps failed and you want to "
            "fix them in isolation. "
            "The steps are given as a space-separated list of: "
            + " ".join(value.value for value in Step)
        ),
        metavar="",
        nargs="+",
        choices=[value.value for value in Step],
    )
    parser.add_argument(
        "--skip",
        help=(
            "If set, skips the specified steps. "
            "This is practical if some of the steps passed and "
            "you want to fix the remainder in isolation. "
            "The steps are given as a space-separated list of: "
            + " ".join(value.value for value in Step)
        ),
        metavar="",
        nargs="+",
        choices=[value.value for value in Step],
    )

    args = parser.parse_args()

    overwrite = bool(args.overwrite)

    selects = (
        [Step(value) for value in args.select]
        if args.select is not None
        else [value for value in Step]
    )
    skips = [Step(value) for value in args.skip] if args.skip is not None else []

    repo_root = pathlib.Path(os.path.realpath(__file__)).parent.parent

    if Step.REFORMAT in selects and Step.REFORMAT not in skips:
        print("Re-formatting...")
        reformat_targets = [
            "bodygamepad",
            "continuous_integration",
            "tests",
            "setup.py",
        ]

        if overwrite:
            exit_code = call_and_report(
                verb="black",
                cmd=["black"] + reformat_targets,
                cwd=repo_root,
            )
            if exit_code != 0:
                return 1
        else:
            exit_code = call_and_report(
                verb="check with black",
                cmd=["black", "--check"] + reformat_targets,
                cwd=repo_root,
            )
            if exit_code != 0:
                return 1
    else:
        print("Skipped re-formatting.")

    if Step.MYPY in selects and Step.MYPY not in skips:
        print("Mypy'ing...")
        mypy_targets = [
            "bodygamepad",
            "tests",
            "continuous_integration",
        ]
        config_file = pathlib.Path("continuous_integration") / "mypy.ini"

        exit_code = call_and_report(
            verb="mypy",
            cmd=["mypy", "--strict", "--config-file", str(config_file)] + mypy_targets,
            cwd=repo_root,
        )
        if exit_code != 0:
            return 1
    else:
        print("Skipped mypy'ing.")

    if Step.PYLINT in selects and Step.PYLINT not in skips:
        print("Pylint'ing...")
        pylint_targets = [
            "bodygamepad",
            "tests",
            "continuous_integration",
        ]
        rcfile = pathlib.Path("continuous_integration") / "pylint.rc"

        exit_code = call_and_report(
            verb="pylint",
            cmd=["pylint", f"--rcfile={rcfile}"] + pylint_targets,
            cwd=repo_root,
        )
        if exit_code != 0:
            return 1
    else:
        print("Skipped pylint'ing.")

    if Step.TEST in selects and Step.TEST not in skips:
        print("Testing...")
        env = os.environ.copy()
        env["ICONTRACT_SLOW"] = "true"

        exit_code = call_and_report(
            verb="execute unit tests",
            cmd=[
                "coverage",
                "run",
                "--source",
                "bodygamepad",
                "-m",
                "unittest",
                "discover",
            ],
            cwd=repo_root,
            env=env,
        )
        if exit_code != 0:
            return 1

        _ = call_and_report(verb="report the coverage", cmd=["coverage", "report"])

        # NOTE (mristin, 2023-03-26):
        # Ignore a return code != 0 since coverage report returns 1 if there are no
        # tests.
    else:
        print("Skipped testing.")

    if Step.DOCTEST in selects and Step.DOCTEST not in skips:
        print("Doctest'ing...")

        doc_files = ["README.rst"]

        exit_code = call_and_report(
            verb="doctest",
            cmd=[sys.executable, "-m", "doctest"] + doc_files,
            cwd=repo_root,
        )
        if exit_code != 0:
            return 1

        for pth in (repo_root / "bodygamepad").glob("**/*.py"):
            if pth.name == "__main__.py":
                continue

            # NOTE (mristin, 2023-03-26):
            # The subprocess calls are expensive, call only if there is an actual
            # doctest
            text = pth.read_text(encoding="utf-8")
            if ">>>" in text:
                exit_code = call_and_report(
                    verb="doctest",
                    cmd=[sys.executable, "-m", "doctest", str(pth)],
                    cwd=repo_root,
                )
                if exit_code != 0:
                    return 1
    else:
        print("Skipped doctest'ing.")

    if (
        Step.CHECK_INIT_AND_SETUP_COINCIDE in selects
        and Step.CHECK_INIT_AND_SETUP_COINCIDE not in skips
    ):
        print("Checking that bodygamepad/__init__.py and setup.py coincide...")
        exit_code = call_and_report(
            verb="check that bodygamepad/__init__.py and setup.py coincide",
            cmd=[
                sys.executable,
                "continuous_integration/check_init_and_setup_coincide.py",
            ],
            cwd=repo_root,
        )
        if exit_code != 0:
            return 1
    else:
        print("Skipped checking that bodygamepad/__init__.py and setup.py coincide.")

    # NOTE (mristin, 2023-04-15):
    # We need to check for the Python version since ``argparse`` output changes
    # between the versions. Hence, we pin it at the moment to Python 3.9.

    if (3, 9) <= sys.version_info < (3, 10):
        if (
            Step.CHECK_HELP_IN_README in selects
            and Step.CHECK_HELP_IN_README not in skips
        ):
            cmd = [sys.executable, "continuous_integration/check_help_in_readme.py"]
            if overwrite:
                cmd.append("--overwrite")

            if not overwrite:
                print("Checking that --help's and the readme coincide...")
                exit_code = call_and_report(
                    verb="check that --help's and the readme coincide",
                    cmd=[
                        sys.executable,
                        "continuous_integration/check_help_in_readme.py",
                    ],
                    cwd=repo_root,
                )
                if exit_code != 0:
                    return 1
            else:
                print("Overwriting the --help's in the readme...")
                exit_code = call_and_report(
                    verb="overwrite the --help's in the readme",
                    cmd=[
                        sys.executable,
                        "continuous_integration/check_help_in_readme.py",
                        "--overwrite",
                    ],
                    cwd=repo_root,
                )
                if exit_code != 0:
                    return 1

            subprocess.check_call(cmd, cwd=str(repo_root))
        else:
            print("Skipped checking that --help's and the doc coincide.")
    else:
        print(
            "Skipped checking that --help's and the doc coincide "
            "since we pin it on Python version 3.9."
        )

    return 0


if __name__ == "__main__":
    sys.exit(main())
